kubeadm: add "+optional" to fields with "omitempty" in v1beta3

neolit123 opened the pull request:
#### What this PR does / why we need it:

Add the "+optional" generator tag to all fields in v1beta3 structs
that have "omitempty" in JSON tags.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubeadm/issues/1727
xref https://github.com/kubernetes/kubeadm/issues/1796

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/970-kubeadm-config
- [KEP]: https://github.com/kubernetes/enhancements/pull/1816
```

^ note that KEP `https://github.com/kubernetes/enhancements/pull/1816` has not merged!

neolit123 committed:
kubeadm: add "+optional" to fields with "omitempty" in v1beta3

Add the "+optional" generator tag to all fields in v1beta3 structs
that have "omitempty" in JSON tags.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -65,6 +69,10 @@ type InitConfiguration struct {
 
+	// Metadata is the standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	Metadata *metav1.ObjectMeta
```
i went with not embedding the structure due to the conflicting ClusterConfiguration.ClusterName and ObjectMeta.ClusterName.

i don't think ObjectMeta.ClusterName is being used in core currently.

the old PR and long discussion was here:
https://github.com/kubernetes/kubernetes/pull/84357

vincepri replied:
Note that embedding ObjectMeta into CRDs usually has implications for validation. IIRC, we removed ObjectMeta from our types and used only the fields needed because ObjectMeta is validated in the api server separately, and the object fails validation due to the custom logic that's in there.

Why do we need this here though?

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:
```diff
@@ -28,15 +28,22 @@ import (
+	// Metadata is the standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	Metadata *metav1.ObjectMeta `json:"metadata,omitempty"`
```
also not embedding it here, which is not consistent with other core and external types.

question for @vincepri @fabriziopandini :

CAPI is not using the stock ObjectMeta since it had a problem around one of its "time" fields:
https://github.com/kubernetes-sigs/cluster-api/blob/3065a926259f682f65cb8331b5f2543b270882a8/api/v1alpha4/common_types.go#L122-L157

what is your recommendation for kubeadm?
one of the original reasons for the ObjectMeta requests for kubeadm were related to kustomize and the ObjecMeta.Name field, which the CAPI object does not have.

vincepri replied:
Ah thanks, just saw this comment after writing the above. I'd not embed it altogether, is there a reason to have it?

neolit123 replied:
frankly, the only requests were from kustomize users for file-on-disk manipulations, which is not a solid enough reason to have the kubeadm types include a partial ObjectMeta or the standard one.

given ObjectMeta is for types that are managed at the apiserver, it feels that we can add it to kubeadm types only once and if they become CRDs.

the discussion on the unmerged KEP brings similar concerns:
https://github.com/kubernetes/enhancements/pull/1816

fabriziopandini replied:
Wondering if someone from API machinery (might be @nikhita) could chime in here; we have two topics at stake
- Should we add ObjectMeta tu kubeadm component config to make them "kustomize friendly" ? 
- If yes, should we add the full ObjectMeta or a subset of it thus avoiding the problems mentioned above ?

neolit123 commented:
/retest

vincepri commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:
```diff
@@ -225,18 +266,23 @@ type BootstrapToken struct {
 	// TTL defines the time to live for this token. Defaults to 24h.
 	// Expires and TTL are mutually exclusive.
+	// +optional
 	TTL *metav1.Duration `json:"ttl,omitempty"`
```
Is 0 a valida TTL in this case? Is there a minimum?

neolit123 replied:
this TTL is used to create the unstuctured format for the bootstrap token v1.Secret type:
https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#bootstrap-token-secret-format

so if the `expiration` is 0s from now, i'd assume that it would pass validation, but the token cleaner controller from core would just GC it on it's next cycle.

fabriziopandini commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -35,6 +35,10 @@ import (
 
+	// Metadata is the standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	Metadata *metav1.ObjectMeta
```
`json:"metadata,omitempty"`
(same below)

fabriziopandini commented:
@neolit123 given the discussion above about the underlying use cases and the technical problems, I'm for bumping the priority of this change to P2 (or P3).

neolit123 commented:
the ObjectMeta commit can be dropped from this PR, if you think the "+optional" change has more value for v1beta3 and projects like Cluster API. at this point, my vote is -1 on adding ObjectMeta in v1beta3 since there is no sufficient evidence for users needing it. i.e. there is no evidence that users are using kustomize for component config like kubelet v1beta1.

neolit123 commented:
- removed the ObjectMeta addition. discussions with @fabriziopandini indicated that this is not really needed currently.
- removed release note as +optional is arguably not user facing for the time being for kubeadm users.

neolit123 commented:
/retest

fabriziopandini commented:
/lgtm
/retest

neolit123 commented:
/hold cancel
